src/mame/f256/tiny_vicky.cpp

```diff
@@ -0,0 +1,613 @@
+// license:BSD-3-Clause
+// copyright-holders:Daniel Tremblay
+#include "emu.h"
+#include "tiny_vicky.h"
+
+#include "machine/ram.h"
+#include "screen.h"
+
+DEFINE_DEVICE_TYPE(TINY_VICKY, tiny_vicky_video_device, "tiny_vicky", "F256K Tiny Vicky")
+
+tiny_vicky_video_device::tiny_vicky_video_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock):
+    device_t(mconfig, type, tag, owner, clock)
+    , m_sof_irq_handler(*this)
+    , m_sol_irq_handler(*this)
+{
+}
+
+tiny_vicky_video_device::tiny_vicky_video_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock):
+    tiny_vicky_video_device(mconfig, TINY_VICKY, tag, owner, clock)
+{
+}
+
+rgb_t tiny_vicky_video_device::get_text_lut(uint8_t color_index, bool fg, bool gamma)
+{
+    uint8_t red =   m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 + 2];
+    uint8_t green = m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 + 1];
+    uint8_t blue =  m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 ];
+    if (gamma)
+    {
+        blue = m_iopage0_ptr[blue];
+        green = m_iopage0_ptr[0x400 + green];
+        red = m_iopage0_ptr[0x800 + red];
+    }
+    return rgb_t(red, green, blue);
+}
+
+rgb_t tiny_vicky_video_device::get_lut_value(uint8_t lut_index, uint8_t pix_val, bool gamma)
+{
+    int lutAddress = 0xD000 - 0xC000 + (lut_index * 256 + pix_val) * 4;
+    if (!gamma)
+    {
+        return rgb_t(m_iopage1_ptr[lutAddress + 2], m_iopage1_ptr[lutAddress + 1], m_iopage1_ptr[lutAddress]);
+    }
+    else
+    {
+        return rgb_t(m_iopage0_ptr[0x800 + m_iopage1_ptr[lutAddress + 2]],
+                    m_iopage0_ptr[0x400 + m_iopage1_ptr[lutAddress + 1]],
+                    m_iopage0_ptr[m_iopage1_ptr[lutAddress]]);
+    }
+}
+
+uint32_t tiny_vicky_video_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+    if (m_running && m_iopage0_ptr)
+    {
```

Comment:
In this function, you really should be making an effort to at least restrict the drawing to the clipping rectangle vertically so that raster effects have a chance of working.